src/libraries/Microsoft.Extensions.Configuration.Binder/gen/ConfigurationBindingSourceGenerator.Helpers.cs

```diff
@@ -0,0 +1,145 @@
+            public const string CollectionNotSupported = "The collection type is not supported";
+            public const string DictionaryKeyNotSupported = "The dictionary key type is not supported";
+            public const string ElementTypeNotSupported = "The collection element type is not supported";
+            public const string MultiDimArraysNotSupported = "Multidimensional arrays are not supported.";
+            public const string NullableUnderlyingTypeNotSupported = "Nullable underlying type is not supported";
+            public const string TypeNotDetectedAsInput = "Generator parser did not detect the type as input";
+            public const string TypeNotSupported = "The type is not supported";
+        }
+
+        private static class TypeFullName
+        {
+            public const string Array = "System.Array";
+            public const string ConfigurationKeyNameAttribute = "Microsoft.Extensions.Configuration.ConfigurationKeyNameAttribute";
+            public const string Dictionary = "System.Collections.Generic.Dictionary`2";
+            public const string GenericIDictionary = "System.Collections.Generic.IDictionary`2";
+            public const string HashSet = "System.Collections.Generic.HashSet`1";
+            public const string ISet = "System.Collections.Generic.ISet`1";
+            public const string IConfigurationSection = "Microsoft.Extensions.Configuration.IConfigurationSection";
+            public const string IConfiguration = "Microsoft.Extensions.Configuration.IConfiguration";
+            public const string IDictionary = "System.Collections.Generic.IDictionary";
+            public const string IServiceCollection = "Microsoft.Extensions.DependencyInjection.IServiceCollection";
+            public const string List = "System.Collections.Generic.List`1";
+        }
+
+        private static bool TypesAreEqual(ITypeSymbol first, ITypeSymbol second)
+                => first.Equals(second, SymbolEqualityComparer.Default);
+
+        private enum InitializationKind
+        {
+            None = 0,
+            SimpleAssignment = 1,
+            AssignmentWithNullCheck = 2,
+            Declaration = 3,
+        }
+
+        private sealed class SourceWriter
+        {
+            private readonly StringBuilder _sb = new();
+            private int _indentationLevel;
+
+            public int Length => _sb.Length;
+            public int IndentationLevel => _indentationLevel;
+
+            public void WriteBlockStart(string declaration)
+            {
+                WriteLine(declaration);
+                WriteLine("{");
+                _indentationLevel++;
+            }
+
+            public void WriteBlockEnd(string? extra = null)
+            {
+                _indentationLevel--;
+                Debug.Assert(_indentationLevel > -1);
+                WriteLine($"}}{extra}");
+            }
+
+            public void WriteLine(string source)
+            {
```

Comment:
Nit: this does not take line breaks in the source string into account. Clearly your emitter code is careful to avoid this, but it would be nice for readability if multi-line expressions could be formatted as follows:

```C#
sourceWriter.WriteLine($$"""
             public static void MyGeneratedMethod({{typeFQN}} value)
             {
                   if (value is not null) Console.WriteLine(value);
             }
             """);
```

(Note that in triple-quote strings, the indentation of the closing `"""` is significant as the compiler ignores the preceding whitespace in all lines)